BulkProcessor retries after request handling has been rejected due to a full thread pool

danielmitterdorfer opened the pull request:
With this commit we introduce limited retries with a backoff logic to BulkProcessor
when a bulk request has been rejeced with an EsRejectedExecutionException.

The handling is encapsulated entirely in BulkProcessor and currently no
configuration parameters are exposed to the client.

Fixes #14620.

danielmitterdorfer commented:
The implementation introduces a (package private) class `Retry` which encapsulates the actual retry and backoff logic. It is intended to be used for all sorts of retries within Elasticsearch based on an arbitrary exception but is currently tied to the client API. This is mainly due to the fact that the current listener-based client API does not compose well.

I have thought whether it makes sense to adapt the client API to be more in line with `CompletableFuture`from JDK 8 which would be easier to compose. However, this is entirely different (and probably huge) topic.

Depending on the review comments it could make sense to turn `Retry` into a proper public class and move it probably to `org.elasticsearch.common.util.concurrent` which would mean that we have to introduce a thin adapter layer for the client API.

I've also extracted `BulkRequestHandler` from `BulkProcessor` as it was already quite complex to understand and stuffing even more logic into the same class felt wrong to me.

tlrx commented on core/src/main/java/org/elasticsearch/action/bulk/Retry.java:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    /**
+     * Invokes #bulk(BulkRequest, ActionListener) on the provided client. Backos off on the provided exception and delegates results to the
```
> Backos off

Typo

tlrx commented:
Not much to say, I like it. Do you think we can add a test like `RejectionActionIT`?

danielmitterdorfer commented:
@tlrx Thanks for your review and your feedback. I'll fix the typo and I already looked at `RejectionActionIT`. I think it is hard to test as the only side effect you would notice is that your requests may slow down due to backoff. As backoff is finite it is still possible we get `EsRejectedExecutionException`, it is just less likely but in a very slow environment the test could lead to false positives.

One thing I have pondered is whether it makes sense to expose retry parameters (number of retries and initial pause time) to users of `BulkProcessor` via its builder at the risk of misconfigurations. This would allow us to raise the number of retries for the test significantly. If we don't want to increase the API surface area, a less intrusive option would be to expose the setters for retry parameters just as package local for the test.

What do you think?

tlrx commented:
@danielmitterdorfer I like the current code but I find it difficult to test or to configure because the retry and backoff logics are obfuscated in the implementation. 

I think we could add a default retry logic that retries on `EsRejectedException` and that can be configured (number of retries / initial delay) via the BulkProcessor builder. In addition, it would be nice to be able to plug its own retry and/or backoff logic. This would make the code more testable.

danielmitterdorfer commented:
I have started to implement an integration test based on your suggestion but I noticed that we do not get `EsRejectedExecutionException` in the callback `onFailure(Throwable)` but in `onResponse(BulkResponse)` by iterating over the results and checking individual items.

Consequently, we cannot just retry the whole operation but must split the bulk request into multiple bulk requests internally.

Consider this scenario for example:

1. First bulk request for ids \[1, 2, 3, 4, 5, 6, 7, 8, 9, 10\] (concrete operation is irrelevant for this example)
2. We get success for [1, 2, 3, 4, 5] but a failure due to `EsRejectedExecutionException` for [6, 7, 8, 9, 10].
3. We backoff a bit and construct a new bulk request for [6, 7, 8, 9, 10].
4. We get success for [6, 7, 8, 9] but a failure due to `EsRejectedExecutionException` for [10].
4. We backoff again and construct a new bulk request for [10].
5. We get success for [10].

At this point, we have to notify the `BulkProcessor.Listener` by invoking its callback methods. I can think of two possibilities:

1. Call the callback method every time with the subset of the responses we got back. This is quite straightforward but we leak retry semantics to the listener which might (quite reasonably) expect that the callback method is just invoked once per bulk request.
2. Collect all responses (i.e. [1, 2, 3, 4, 5], [6, 7, 8, 9] and [10]) and construct a new synthetic bulk response and invoke the listener only once.

Although option 2 is more complex to implement, we do not leak retry semantics to `BulkProcessor.Listener`, hence I tend to prefer this approach. A detail in this case: I'd implement `BulkResponse.tookInMillis` in wall clock time semantics, i.e. sum of all BulkResponse.tookInMillis + sum of all backoff times. To me this is the most sensible option from a user perspective.

One additional thing that bothers me that it is possible that we invoke the bulk API with as little as one element but I think working around this problem in the async case we would introduce much more complex code and in the sync case it is not possible at all.

wdyt @tlrx? /cc: @bleskes

bleskes commented:
I think of the bulk processor as an abstraction around the bulk API, allowing the user to work with single item requests and have the bulk done automatically behind the scenes. As such I don't think we need to worry about maintaining internal bundling of requests. These are not controlled by the user explicitly anyway.

I would go with the simplest implementation and throw the reject items back into the pending request list, while blocking outgoing requests for a given period of time. They will be executed when the block is removed. (i.e., option #1)

Last - looking at the code I see that our listener is bulk based:

```
    void afterBulk(long executionId, BulkRequest request, BulkResponse response);
```

I think it would be good if the listener also have methods that work on the item level - so people really wouldn't have to work with the bulks at all. You put item request in and you get iterm requests out. Whether we want to do it as part of the listener API or as a basic abstract listener implementation is a good question.

Last I wonder - what is the role of the execution ID in that listener - can you come up with a use case for it?

tlrx commented:
I agree with @bleskes suggestions :)

danielmitterdorfer commented:
@bleskes, @tlrx: Thanks for your feedback.

Regarding not maintaining bundling: that's good. :)

Regarding the signature `void afterBulk(long executionId, BulkRequest request, BulkResponse response)`. I like your idea of working just on item level.

But the interface is already there, so we can do either:

* Break backwards compatibility and change it the way you suggested (I also think it is better to get rid of `BulkResponse`in the listener API). I think this is just an option if we don't backport to 2.x..
* Introducing an abstract base class in addition to the interface with the new methods and deprecating the interface

I have no clue what the execution id is for, except for some kind of request correlation between a `beforeBulk` operation and an `afterBulk` operation. Otherwise the client would not know that (although I am not sure whether knowing that is relevant...).

So here is what I would do based on your suggestions:

1. Reissue the request internally on any failed items but failing immediately if any of the failures is not of type `EsRejectedExecutionException`.
2. (Pending on your feedback) introduce three new operations based on individual items in a new interface `BulkProcessor.ItemListener`, removing `BulkProcessor.Listener` in master and deprecating it in 2.x. This way we can introduce it without needing clients to change in 2.x. The interface looks like:

```
void beforeExecute(IndexRequest request);
void afterExecute(IndexRequest request, ActionWriteResponse response);
void afterExecute(IndexRequest request, Throwable failure);
```

bleskes commented:
> except for some kind of request correlation between a beforeBulk operation and an afterBulk operation. Otherwise the client would not know that (although I am not sure whether knowing that is relevant...).

That's actually a valid thing. Imho it shouldn't be part of the listener but rather protected methods. if people want to override and do something custom they can do it by sub classing the class.

tlrx commented:
> Reissue the request internally on any failed items but failing immediately if any of the failures is not of type EsRejectedExecutionException.

Just to be sure that I correctly understand this point: only rejected item requests should be re-added to the bulk processor. Other failures should directly call `afterExecute(IndexRequest request, ...)`

> Imho it shouldn't be part of the listener but rather protected methods. if people want to override and do something custom they can do it by sub classing the class.

+1

danielmitterdorfer commented:
> Just to be sure that I correctly understand this point: only rejected item requests should be re-added to the bulk processor. Other failures should directly call `afterExecute(IndexRequest request, ...)`

The exact behavior is:

* If and only if the bulk response consists of only items with success, we call the listener's `onResponse` callback.
* If and only if the bulk response consists of either items with success or items with failure == EsRejectedExecutionException, we will reissue another bulk request for the failed items.
* Otherwise, we fail immediately and delegate either to the listener's `onResponse` callback or `onFailure` callback (depending on which callback method has been called for our internal listener).

bleskes commented:
If we move to item level based listener (+1), we can call the listener on every successful item. failed item will be checked for a rejection, if so it will be put in the queue. Otherwise call the on failure method of the listener. Makes sense?

danielmitterdorfer commented:
@bleskes, @tlrx: Thanks for the hints. Could one of you look at the new version?

The implementation was more straightforward and self-contained by keeping the current listener API. Hence, I suggest we keep the listener API for this PR and open a new one for just changing the listener API as this is a good self-contained change, I think.

A short summary of the changes:

* The backoff logic checks now `BulkItemResponse#isFailed()` and its related failure instead of catching an actual exception. Failed items are collected and a request is reissued according to the backoff policy.
* The default backoff policy progresses now exponentially instead of geometrically. Its default configuration leads to the progression: 50ms, 60ms, 80ms, 150ms, 280ms, 580ms, 1.2s, 2.7s.
* The backoff policy can be easily changed by users of `BulkProcessor` via its builder; it is also possible to provide custom policies. The default is to have no backoff (behavior as it is now). I'll update the reference documentation in a separate PR.
* I've added `BulkProcessorRetryIT` which contrasts a bulk request with and without backoff. We check in the backoff case that we do not accidentally introduce duplicates by retrying. 

One thing that bothers me and that I am not really sure about, is that I had to add a mock client for a unit test (unfortunately, retrying requests requires having access to the client).

As we do not use a mocking framework in core, I have added an "empty" implementation of `Client` in test-framework (`MockClient`). I have mixed feelings about mocking frameworks and I rather use mocking sparingly, but I sense this would be one of the cases where it  makes sense (but the result of #7489 indicates that the problems outweigh the benefits so I am fine with that).

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/BackoffPolicy.java:
```diff
@@ -0,0 +1,141 @@
+    private static class ExponentialBackoffIterator implements Iterator<TimeValue> {
+        private final int numberOfElements;
+
+        private int start;
```
final?

danielmitterdorfer replied:
Sure, I totally missed it. Thanks for pointing it out.

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/BackoffPolicy.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    /**
+     * Creates an new exponential backoff policy with a default configuration of 50 ms initial wait period and 8 retries.
```
You may want to put the total length of all retries in the javadoc. Its an interesting number to think about.

danielmitterdorfer replied:
I am not sure what you mean by "total length of all retries". Do you refer to the progression: 50ms, 60ms, 80ms, 150ms, 280ms, 580ms, 1.2s, 2.7s. or to the total wait time, which is 5.1s?

tbh, I think this is an implementation detail and it might be easy to forget updating the documentation in case the backoff progression function is changed.

nik9000 replied:
5.1s. Abstractions that can have a profound effect on runtime should leak a little. I agree it'd be easy to forget to update the docs. Maybe just a hint that it is around five seconds. From the docs now its hard to infer how high it goes. Every time I see exponential backoff my instinct is for times that go up into the minutes. Which isn't right here.

danielmitterdorfer replied:
Done.

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/BulkRequestHandler.java:
```diff
@@ -0,0 +1,150 @@
+            } catch (Exception e) {
+                if (!afterCalled) {
+                    listener.afterBulk(executionId, bulkRequest, e);
+                }
```
Is it right to just eat the exception thrown from the listener? At least log a warning or something.

danielmitterdorfer replied:
I have just moved code around, so this implementation is not new. I also think it is ok here not to do anything because the intention is that the listener handles the exception and we just pass it on. But if you still want to, I am not reluctant to log a warning in addition here.

nik9000 replied:
> I have just moved code around, so this implementation is not new.

Fair enough. I'd still log a warning just to help debug any mistakes in the listener. If all goes well and the listener catches any exceptions then we will never call it.

danielmitterdorfer replied:
I issue a warning now.

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/Retry.java:
```diff
@@ -0,0 +1,226 @@
+        private final Client client;
+        private final ActionListener<BulkResponse> listener;
+        private final Iterator<TimeValue> backoff;
+        private final ScheduledExecutorService s = new ScheduledThreadPoolExecutor(1);
```
Should we be spawning a new thread here? Other places ES is pretty careful to use some pool.

danielmitterdorfer replied:
You're right. This is the only place in the code base I could find that is using a `ScheduledExecutorService` directly. However, as `BulkProcessor` is intended as a client-side API I think it *could* make sense to do it that way. Otherwise, the API surface area is increasing as clients have to create an instance of `ThreadPool` now and pass it to `BulkProcessor`. wdyt?

nik9000 replied:
So I'm thinking of using this in the reindex API which'd make it used inside of Elasticsearch. Taking a ThreadPool is fairly normal for the internals of Elasticsearch. I suppose if you wanted to keep API backwards compatibility then you could make it optional and the whole thing to fail if one isn't provided but a backoff needs it.

danielmitterdorfer replied:
I have found a much easier alternative: As we interact via the client with the cluster, the most natural thing is to have the client schedule the next request after the backoff period.

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/Retry.java:
```diff
@@ -0,0 +1,226 @@
+            int index = 0;
+            for (BulkItemResponse bulkItemResponse : bulkItemResponses.getItems()) {
+                if (bulkItemResponse.isFailed()) {
+                    // Assumption: ES maintains order between request and response
```
s/Assumption: //

Its part of the contract for the bulk api.

danielmitterdorfer replied:
I have removed the comment.

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/Retry.java:
```diff
@@ -0,0 +1,226 @@
+        private void retry(BulkRequest bulkRequestForRetry) {
+            if (backoff.hasNext()) {
+                s.schedule((() -> this.execute(bulkRequestForRetry)), backoff.next().millis(), TimeUnit.MILLISECONDS);
+            }
```
Does the listener get notified if we run out of retries?

nik9000 replied:
Ok - I see where it is called. These checks are a bit too distant for my taste.

danielmitterdorfer replied:
Well, the main entry point for all of the handling logic is in `AbstractRetryHandler#onResponse(BulkResponse)`. You just happened to look first in `#retry(BulkRequest)`. However, `backoff.hasNext() == true` is actually a precondition for this method and I think we can remove the check here entirely for two reasons:

1. It is already checked in `#canRetry()`
2. This method is private and only ever called from a single thread so there is no need to recheck.

wdyt?

nik9000 replied:
> This method is private and only ever called from a single thread so there is no need to recheck.

I'm just weary of having the failure handling case so far from the success case. I figure its harder for someone to break it if its closer together.

danielmitterdorfer replied:
The more we talk about it, the more I sense we're talking about different things. :)

I think all the code is quite well encapsulated in `AbstractRetryHandler#onResponse()` (I've annotated it with a few comments below):

```java
public void onResponse(BulkResponse bulkItemResponses) {
            if (!bulkItemResponses.hasFailures()) {
                // Handle success case - just return all responses gathered so far
                addResponses(bulkItemResponses, (r -> true));
                finishHim();
            } else {
                // We've got some kind of failure in at least one of the item responses.
                if (canRetry(bulkItemResponses)) {
                   // We are able to retry, so retry all failed reponses and memorize all that were ok...
                    addResponses(bulkItemResponses, (r -> !r.isFailed()));
                    retry(createBulkRequestForRetry(bulkItemResponses));
                } else {
                    // ... we cannot retry because we either the backoff policy doesn't let us or we got at least one failure which is not considered retriable. Bail out.
                    addResponses(bulkItemResponses, (r -> true));
                    finishHim();
                }
            }
        }
```

The rest is just implementation details in the individual methods that are called.

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/Retry.java:
```diff
@@ -0,0 +1,226 @@
+                } else {
+                    addResponses(bulkItemResponses, (r -> true));
+                    BulkResponse aggregatedResults = getCompleteResponse();
+                    notifyListener(aggregatedResults);
```
Maybe drop the argument to `notifyListener` its always going to return the aggregatedResults, right?

danielmitterdorfer replied:
You're right. This makes the code simpler.

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/Retry.java:
```diff
@@ -0,0 +1,226 @@
+            }
+        }
+
+        private BulkResponse getCompleteResponse() {
```
`synthesizeResponse`? I just want something to make it obvious that it the result of squashing together lots of responses.

danielmitterdorfer replied:
Another idea: `getAccumulatedResponse()`?

nik9000 replied:
Fine by me. I tend not to use `get` unless I'm implementing a method to get a property just because getters are so common in java. If I ever think of the method as building a thing out of state  then I don't use `get`. Its kind of silly because the whole point of getters is that you can build the thing. `getAccumulatedResponse` is fine.

danielmitterdorfer replied:
Not using the `get` prefix takes some getting used to for me. Thanks for explaining your reasoning. I've named the method now `getAccumulatedResponse` as per your suggestion.

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/Retry.java:
```diff
@@ -0,0 +1,226 @@
+        private final Iterator<TimeValue> backoff;
+        private final ScheduledExecutorService s = new ScheduledThreadPoolExecutor(1);
+        private final Class<? extends Throwable> retryOnThrowable;
+        private final List<BulkItemResponse> responses = new ArrayList<>();
```
Maybe a comment about why you don't need a special list to satisfy visibility. I think this is ok as is because there is just one thread in the thread pool but if you move this to a generic thread pool you might have to jump through more hoops to make the visibility guarantees line up.

danielmitterdorfer replied:
You're right; this is rather subtle and the more I think about it, the more uneasy I am about the current state. I'd rather play safe here and use client-side locking on all list accesses, wdyt?

nik9000 replied:
> I'd rather play safe here and use client-side locking on all list accesses in order to avoid that another change introduces a subtle visibility issue

I always get a bit jumpy about just locking things because I don't have a great understanding of when Java is allowed to reorder instructions around locks vs volatile. Its worth looking into more.

danielmitterdorfer replied:
I think we have basically two options here:

1. Using `CopyOnWriteArrayList`: This class is intended for seldom updates and lots of writes as it copies the underlying array on every write. I assume the code path here is not that performance-critical but I tend not to use `CopyOnWriteArrayList` here as it does not seem to be a good fit for the intended usage pattern.
2. Synchronize all accesses to the collection, which are in `#addResponses()` and `#getAccumulatedResponse()`. Apart from visibility across threads, we need to ensure adding items to this collection and copying it is atomic. So synchronizing is just the right thing to do here (`volatile` would just buy us visibility). We could get more fancy by using other lock implementations but I think that's overkill in this situation. I have not measured it but for this usage pattern I would not expect any lock contention.

As you've also talked about reordering: The JMM effectively requires JVM implementations to issue a memory barrier for volatile accesses and also for synchronized blocks to establish a happens-before relation between memory accesses.

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/BackoffPolicy.java:
```diff
@@ -0,0 +1,141 @@
+    /**
+     * Creates an new exponential backoff policy with a default configuration of 50 ms initial wait period and 8 retries.
+     *
+     * @return A backoff policy with an exponential increase in wait time for retries. The returned instance is thread safe.
```
But the Iterators it returns aren't.

danielmitterdorfer replied:
I think (a) that's kinda obvious and (b) the only component that should use such an iterator is the bulk processor but I've added it to the Javadocs now.

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/Retry.java:
```diff
@@ -0,0 +1,226 @@
+            try {
+                listener.onResponse(response);
+            } finally {
+                s.shutdownNow();
```
I'm not sure why, but the usual convention for freeing resources when the request is done is to have a method called `finishHim`. Mortal combat reference? Anyway, the nice thing about this convention is that it gives us a place to look for resources to be freed. But above I mention reusing a thread pool of some sort anyway.

danielmitterdorfer replied:
Didn't know about that. :) I have changed the name now.

danielmitterdorfer commented:
@nik9000 Thanks for your review; that was very helpful. I have updated the code and asked a few follow up questions. Could you please look at them? Thanks!

nik9000 commented:
> Could you please answer them? Thanks!

Done!

danielmitterdorfer commented:
@nik9000: I've pushed another commit incorporating your comments.

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/BackoffPolicy.java:
```diff
@@ -0,0 +1,143 @@
+     * roughly 5.1 seconds in total.
+     *
+     * @return A backoff policy with an exponential increase in wait time for retries. The returned instance is thread safe but each
+     * iterator created from it, should only be used by a single thread.
```
Nit: drop the `,` here.

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/BackoffPolicy.java:
```diff
@@ -0,0 +1,143 @@
+     */
+    public static BackoffPolicy exponentialBackoff(TimeValue initialDelay, int maxNumberOfRetries) {
+        // cap at Integer.MAX_VALUE, it does not make sense to wait for such a long time period anyway
+        int initialDelayMillis = initialDelay.millis() > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) initialDelay.millis();
```
Just throw an IllegalArgumentException?

danielmitterdorfer commented:
@nik9000 I've changed the Javadoc and throw now an `IllegalArgumentException` instead of silently capping the value. Can you have a (final?) look please?

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/BulkRequestHandler.java:
```diff
@@ -0,0 +1,160 @@
+                                }
+                            }
+                        });
+                success = true;
```
Can you remove `success` entirely and just release if acquired? No need to release in the listener if we're blocking here.

I think it might be worth a comment that this code is running synchronously. A quick scan of the code makes it look like it should be async and you have to read more closely to see the sync.

danielmitterdorfer replied:
Actually, it is required to not release the semaphore until after we're completely done. Otherwise, we might break the limit of maximum allowed concurrent requests the user has defined. We could maybe release the semaphore before we delegate to the listener (line 121 and line 131). Consider this execution trace:

maximum allowed number of concurrent requests: 2

1. A bulk request B0 comes in, we acquire the semaphore. Allowed concurrent requests = 1
2. we issue the request but hit a `EsRejectedExecutionException`. We back off for - say - 50 ms
3. A second bulk request B1 comes in, we acquire the semaphore. Allowed concurrent requests = 0
4. Another bulk request B2 comes in, we acquire the semaphore and we are now blocked as allowed concurrent requests = 0.
5. The first retry request on B0 is sent, returns and we're done. We call `onSuccess` on the callback listener in `AsyncBulkRequestHandler`which in turn will notify its delegate listener. Now we release the semaphore. Post condition: Number of allowed concurrent requests = 1.
6. B2 can now continue.

If we get rid of the `success` check we would release the semaphore immediately after the first attempt to send a bulk request and B2 would break the limit on the number of maximum allowed concurrent requests in step 4 in the above scenario.

@nik9000: Does this answer your remark?

nik9000 replied:
Ok - I've looked again and it looks like this is indeed async. When is success false? Request validation failure? Maybe rename `success` so the next person isn't confused by it. `listenedWillRelease` or something?

nik9000 replied:
`success` just made me think "the whole operation is successful" not "successfully started operation".

danielmitterdorfer replied:
tbh, I have also just moved this code as is. But I try to come up with a better name.

danielmitterdorfer replied:
I've called it `bulkRequestSetupSuccessful` now, indicating that it is just preparing everything.

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/BackoffPolicy.java:
```diff
@@ -0,0 +1,145 @@
+        @Override
+        public TimeValue next() {
+            if (!hasNext()) {
+                throw new NoSuchElementException("Only up to " + numberOfElements + " elements");
```
Nit: `"Only up to [" + numberOfElements + "] elements"` is more normal for elasticsearch. We love putting those values in square brackets!

danielmitterdorfer replied:
You're right. I am not yet used to it. ;) Bear with me; I'll change that.

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/BackoffPolicy.java:
```diff
@@ -0,0 +1,145 @@
+            if (!hasNext()) {
+                throw new NoSuchElementException("Only up to " + numberOfElements + " elements");
+            }
+            int result = start + 10 * ((int) Math.exp(0.8d * (currentlyConsumed)) - 1);
```
`((int) Math.exp(0.8d * currentlyConsumed) - 1)`?

danielmitterdorfer replied:
I am not quite sure what your question is. :) But I assume you wonder about the `start + 10*` part. The constant factor is "expectation management". If the user specifies a start interval of X ms, they expect that the first backoff time is X ms and that's why the "start" part is there. The 10x factor is just to amplify the increase in backoff time. I could also use another factor instead of 0.8...

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/Retry.java:
```diff
@@ -0,0 +1,233 @@
+        return new Retry(retryOnThrowable);
+    }
+
+    // non-standard setter for a fluent API
```
These are so common in elasticsearch its not worth mentioning.

danielmitterdorfer replied:
I'll remove the comment..

nik9000 commented on core/src/main/java/org/elasticsearch/action/bulk/Retry.java:
```diff
@@ -0,0 +1,233 @@
+
+        private void retry(BulkRequest bulkRequestForRetry) {
+            assert backoff.hasNext();
+            scheduledRequestFuture = client.threadPool().schedule(backoff.next(), ThreadPool.Names.BULK, (() -> this.execute(bulkRequestForRetry)));
```
I guess `BULK` is the right thing here. Usually BULK is used on the receiving side but these are the same in spirit as its just coordinating things.

danielmitterdorfer replied:
Agreed.

bleskes replied:
We typically do this light weight coordination on the same thread. I.e., Names.SAME . This does nothingother  than spawn another bulk request. This will cause a new thread to be spawned as we don't do anything else with the bulk pool on the client.  To be honest, I don't think the transport client should have so many thread pools. I'll open a different issue for that.

danielmitterdorfer replied:
Thanks for pointing that out. I changed it.

nik9000 commented on core/src/test/java/org/elasticsearch/action/bulk/BulkProcessorRetryIT.java:
```diff
@@ -0,0 +1,193 @@
+            } else {
+                Throwable t = (Throwable) response;
+                // we're not expecting any other errors
+                fail("Got unexpected failure: " + t);
```
Throwable's toString eats stack traces. I'd just `throw new RuntimeException("unexpected failure", e);`

danielmitterdorfer replied:
Touche. That's not nice. I'll change that.

nik9000 commented on core/src/test/java/org/elasticsearch/action/bulk/BulkProcessorRetryIT.java:
```diff
@@ -0,0 +1,193 @@
+                        Throwable rootCause = ExceptionsHelper.unwrapCause(failure.getCause());
+                        // we're overwhelming the pool, that's the exception we want to get
+                        if (rootCause instanceof EsRejectedExecutionException == false) {
+                            fail("Got unexpected reason [" + failure.getMessage() + "]");
```
Same as below.

nik9000 replied:
One day `getMessage` will be added to the forbidden apis. May as well not make it worse.

danielmitterdorfer replied:
Well, this is not calling Exception#getMessage() but rather Failure#getMessage() so I guess this will not go into forbidden APIs. And apart from that: It is a test, I know that it is an `EsRejectedExecutionException` and I am just interested in the message...

nik9000 commented on core/src/test/java/org/elasticsearch/action/bulk/BulkProcessorRetryIT.java:
```diff
@@ -0,0 +1,193 @@
+                        Throwable rootCause = ExceptionsHelper.unwrapCause(failure.getCause());
+                        // we're not expecting that we still overwhelmed it even once
+                        if (rootCause instanceof EsRejectedExecutionException) {
+                            StringWriter sw = new StringWriter();
```
Just wrap and rethrow and let junit report the exception.

danielmitterdorfer replied:
Aside: I prefer an explicit call to `#fail` to throwing an exception but I see your point that information is lost. I wonder why JUnit does not provide a version of `#fail()` taking a `Throwable` parameter.

nik9000 commented on core/src/test/java/org/elasticsearch/action/bulk/BulkProcessorRetryIT.java:
```diff
@@ -0,0 +1,193 @@
+        }
+
+        // ensure that we can search for the new documents
+        client().admin().indices().prepareRefresh().execute().actionGet();
```
`refresh();`

nik9000 replied:
This is common enough that there is no real need to comment it.

nik9000 replied:
Also, `.get()` is much more common than `.execute().actionGet()`.

danielmitterdorfer replied:
Thanks for pointing it out. I'll change that.

nik9000 commented on core/src/test/java/org/elasticsearch/action/bulk/BulkProcessorRetryIT.java:
```diff
@@ -0,0 +1,193 @@
+
+    private static void indexDocs(BulkProcessor processor, int numDocs) {
+        for (int i = 1; i <= numDocs; i++) {
+            processor.add(new IndexRequest("test", "test", Integer.toString(i)).source("field", randomRealisticUnicodeOfLengthBetween(1, 30)));
```
`client().prepareIndex("test", "test", ....` is more common from what I've seen.

Not for this PR: it might be nice to have a `randomDoc(String index, String type, int id)` method that makes a random document. We use this `Integer.toString` idiom so so so much in the tests. The field being a realistic unicode string is a good touch too.

danielmitterdorfer replied:
I can create an improvement ticket for that.

nik9000 commented on core/src/test/java/org/elasticsearch/action/bulk/BulkProcessorRetryIT.java:
```diff
@@ -0,0 +1,193 @@
+        }
+    }
+
+    public void testSimulatedBulkRejectionLoadWithBackoff() throws Throwable {
```
So far as I can tell this doesn't simulate anything - it causes the rejections outright. So `testBulkRejectionLoadWithBackoff`?

Also, can these two test cases share any code? They are totally two sides of the same coin but I think we're less likely to introduce errors if you pull some of the common code into methods they can call. Lots of other Elastic reviewers won't comment on this but I've waded through lots of copy and pasted test code so I'm extra sensitive to it.

danielmitterdorfer replied:
Normally I am also sensitive to this. How could I not see that? I've pulled out an internal method that is doing everything now. The actual test methods just specify the backoff policy and define whether encountering an `EsRejectedExcecutionException` is ok.

nik9000 commented on core/src/test/java/org/elasticsearch/action/bulk/RetryTests.java:
```diff
@@ -0,0 +1,245 @@
+    }
+
+
+    private static final class ConstantBackoff extends BackoffPolicy {
```
Maybe pull this guy into BackoffPolicy? I figure it'd be useful for anyone uses BulkProcessor in tests. You'd want to make the DELAY configurable, but that is pretty simple.

danielmitterdorfer replied:
Nice catch. I can do that.

nik9000 commented:
I've left another batch of comments. There isn't anything left that should block merging. Address whichever comments you think are important and merge when you are ready. No need to wait for me to review again. If we were in the same time zone it'd be worth it but we're way too many hours off.

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java:
```diff
@@ -79,6 +78,7 @@
         private int bulkActions = 1000;
         private ByteSizeValue bulkSize = new ByteSizeValue(5, ByteSizeUnit.MB);
         private TimeValue flushInterval = null;
+        private BackoffPolicy backoffPolicy = BackoffPolicy.noBackoff();
```
I know I'm late to the party, but the point of #14620 was to have a default that does back off and retry. Why was a no back off chosen as a default?

danielmitterdorfer replied:
To me, "default listener implementation" just meant that we need an implementation that is capable of doing that without having the user write one. The behavior of not backing off was chosen because it was the behavior up to now and the user can easily change it via a new builder method (user docs follow soon). 

But I can change the default behavior to exponential backoff quite simply. It's a one line change (apart from updating Javadocs appropriately).

bleskes replied:
yeah, I meant that we do the right thing and back off by default. Can you please change it? sorry for being late..

danielmitterdorfer replied:
Sure, no problem. I'll do that soon.
